ClientConnectionTest.destroyConnection_whenDestroyedMultipleTimes_thenListenerRemoveCalledOnce

```
java.lang.AssertionError: connection removed should be called only once expected:<1> but was:<2>
    at org.junit.Assert.fail(Assert.java:88)
    at org.junit.Assert.failNotEquals(Assert.java:834)
    at org.junit.Assert.assertEquals(Assert.java:645)
    at com.hazelcast.client.ClientConnectionTest.destroyConnection_whenDestroyedMultipleTimes_thenListenerRemoveCalledOnce(ClientConnectionTest.java:139)
```

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x-IbmJDK1.8/com.hazelcast$hazelcast-client/310/testReport/junit/com.hazelcast.client/ClientConnectionTest/destroyConnection_whenDestroyedMultipleTimes_thenListenerRemoveCalledOnce/

--- a/hazelcast-client/src/main/java/com/hazelcast/client/connection/nio/ClientConnectionManagerImpl.java
+++ b/hazelcast-client/src/main/java/com/hazelcast/client/connection/nio/ClientConnectionManagerImpl.java
@@ -367,11 +367,8 @@ public class ClientConnectionManagerImpl implements ClientConnectionManager {
     @Override
     public void destroyConnection(final Connection connection, final String reason, final Throwable cause) {
         Address endpoint = connection.getEndPoint();
-        if (endpoint != null) {
-            final ClientConnection conn = connections.remove(endpoint);
-            if (conn == null) {
-                return;
-            }
+        ClientConnection conn = (ClientConnection) connection;
+        if (endpoint != null && connections.remove(endpoint, conn)) {
             conn.close(reason, cause);
             for (ConnectionListener connectionListener : connectionListeners) {
                 connectionListener.connectionRemoved(conn);

--- a/hazelcast-client/src/test/java/com/hazelcast/client/ClientConnectionTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/ClientConnectionTest.java
@@ -25,6 +25,8 @@ import com.hazelcast.client.test.TestHazelcastFactory;
 import com.hazelcast.core.Client;
 import com.hazelcast.core.HazelcastInstance;
 import com.hazelcast.core.IExecutorService;
+import com.hazelcast.core.LifecycleEvent;
+import com.hazelcast.core.LifecycleListener;
 import com.hazelcast.core.Member;
 import com.hazelcast.nio.Address;
 import com.hazelcast.nio.Connection;
@@ -48,12 +50,14 @@ import java.util.concurrent.Callable;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicReference;
 
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
 
 @RunWith(HazelcastParallelClassRunner.class)
 @Category({QuickTest.class, ParallelTest.class})
@@ -133,7 +137,26 @@ public class ClientConnectionTest extends HazelcastTestSupport {
         final Address serverAddress = new Address(server.getCluster().getLocalMember().getSocketAddress());
         final Connection connectionToServer = connectionManager.getConnection(serverAddress);
 
+        final CountDownLatch isConnected = new CountDownLatch(1);
+        clientImpl.getLifecycleService().addLifecycleListener(new LifecycleListener() {
+            @Override
+            public void stateChanged(LifecycleEvent event) {
+                if (LifecycleEvent.LifecycleState.CLIENT_CONNECTED == event.getState()) {
+                    isConnected.countDown();
+                }
+            }
+        });
+
         connectionManager.destroyConnection(connectionToServer, null, null);
+
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run()
+                    throws Exception {
+                assertTrue(isConnected.await(5, TimeUnit.SECONDS));
+            }
+        });
+
         connectionManager.destroyConnection(connectionToServer, null, null);
 
         assertEquals("connection removed should be called only once", 1, listener.count.get());

